### ISSUE TITLE
Support win32

### ISSUE THREAD

**Issue by wolvever** [confluentinc/confluent-kafka-python#132]
Now that librdkafka 0.9.4 RC1 support compiling to win32 very easily.  And a .NET version kafka SDK is out there.  It would be nice if confluent-kafka-python adds support compiling on windows.  

I've compiled  librdkafka 0.9.4 RC1 locally.  I'm now stuck at pip install confluent-kafka-python-0.9.4-RC1 source with it.  Python 3.6 + Visual Studio 2015 solve missing inttypes.h problems.  And I'm stuck at link phrase.  The python library requires static library file (rdkafka.lib), if I change librdkafka project to produce lib file.  Then the pip install stops at the following error:

    Creating library build\temp.win-amd64-3.6\Release\confluent_kafka/src\cimpl.cp36-win_amd64.lib and object build\temp.win-amd64-3.6\Release\confluent_kafka/src\cimpl.cp36-win_amd64.exp
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_topic_conf_set_opaque
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_topic_conf_destroy
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_topic_partition_list_add
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_conf_new
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_msg_partitioner_consistent_random
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_conf_set_stats_cb
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_msg_partitioner_random
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_get_err_descs
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_topic_conf_set_partitioner_cb
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_msg_partitioner_consistent
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_topic_partition_list_destroy
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_version
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_version_str
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_conf_set_error_cb
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_topic_conf_new
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_conf_set_opaque
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_err2name
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_topic_conf_set
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_err2str
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_conf_set_default_topic_conf
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_message_timestamp
    confluent_kafka.obj : error LNK2001: unresolved external symbol strcasecmp
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_yield
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_topic_name
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_conf_set
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_topic_partition_list_new
    confluent_kafka.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_conf_destroy
    Producer.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_poll
    Producer.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_destroy
    Producer.obj : error LNK2001: unresolved external symbol __imp_rd_kafka_last_error

Enviornment:
Windows 7 & Visual Studio Community 2015
python 3.6 anaconda 4.3.0
librdkafka 0.9.4 RC1
confluent-kafka-python-0.9.4-RC1 / confluent-kafka-python-0.9.2(the same error)

**Comment by edenhill** [comment]
Proper Windows support for our Python client is on the roadmap.
To avoid dependency problems we will probably provide a pre-built binary Python package including librdkafka, we are thinking of the same for OSX as well (linux is less of an issue with standard dependency management through rpm and deb).

I don't have a time frame for you but feedback does affect the backlog.

**Comment by Data-drone** [comment]
any updates?

**Comment by bl0rq** [comment]
I got one step further w/ this by adding this to an .h file to get around the strcasecmp.  
`#if defined(_WIN32) || defined(_WIN64) 
  #define snprintf _snprintf 
  #define vsnprintf _vsnprintf 
  #define strcasecmp _stricmp 
  #define strncasecmp _strnicmp 
#endif`

This got the build to build, but the example is failing to load the dll at runtime.  (Total python newb, so very well might be EBKAC)

**Comment by edenhill** [comment]
Thanks for looking into this @bl0rq 
I had a stab at AppVeyor Windows builds last week but didn't even get that far due to the default measly Python toolchain on Windows which lacked stuff like stdint.h.
https://ci.appveyor.com/project/ConfluentClientEngineering/confluent-kafka-python/build/1.0.17

**Comment by bl0rq** [comment]
Ah yes, I did a bit of a hackjob there... I took the stdint.h and inttypes.h from this google code archive and stuck them in the include folder.  https://code.google.com/archive/p/msinttypes/downloads  

That got things to build.  Talking w/ python dev, I was able to figure out the runtime error.  It was looking for referenced DLLs that were not around.  So I pulled the librdkafka and deps from the nuget package our C# app uses ( librdkafka.redist from https://www.nuget.org/packages/librdkafka.redist/ ) and it seems to be working!

**Comment by edenhill** [comment]
@bl0rq That's awesome!

Out of curiosity, are you using the standard Python environment for Windows, or the Visual Studio one, or something else?
Are you successfully using `pip install ..` with other packages?

**Comment by bl0rq** [comment]
Good questions!  Python is WinPython-64bit-2.7.10.3.  I installed the full Windows SDK 7.0 to get the VC++ 9 compiler from https://www.microsoft.com/download/details.aspx?id=3138.  (There is a standalone specifically for python but it was blocked "by my administrator" for reasons I did NOT want to dig into lol)  That plus adding .h files (including above redirect of strcasecmp) then pip install was able to do its magic.

**Comment by edenhill** [comment]
Thank you, that's very valuable information!

**Comment by ALTAMASH93** [comment]
@bl0rq @edenhill I managed to setup confluent_kafka on windows, but while importing its failing with -
>>> import confluent_kafka
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "build\bdist.win-amd64\egg\confluent_kafka\__init__.py", line 2, in <module>
  File "build\bdist.win-amd64\egg\confluent_kafka\cimpl.py", line 7, in <module>
  File "build\bdist.win-amd64\egg\confluent_kafka\cimpl.py", line 6, in __bootstrap__
ImportError: DLL load failed: The specified procedure could not be found.

Any thoughts how this can be fixed?

**Comment by edenhill** [comment]
@ALTAMASH93 Did you try the new experimental windows wheels?
https://ci.appveyor.com/project/ConfluentClientEngineering/confluent-kafka-python/build/0.11.0.184/artifacts

Download and `pip install ..` the wheel that matches your platform.

**Comment by ALTAMASH93** [comment]
@edenhill hey thanks! got it working, issue was missing msvcr90.dll.

**Comment by anvithap123** [comment]
@ALTAMASH93 Hi, our team is also trying to use confluent-kafka-python on windows. We are struggling to get this setup. Could you please share us the steps of how you got confluent-kafka-python installed on windows?

**Comment by ehegenberger** [comment]
I'm also trying to get this to work, any progress? Failing that, does anyone know if confluent-kafka-python will work in the Windows 10 subsystem for linux?

**Comment by olejorgenb** [comment]
> Proper Windows support for our Python client is on the roadmap. (2017)

As much as I love to push my coworkers onto linux it would be nice with a timeline indication :) We can manage a while, but if the timeline line is 6months+ we need to look for workarounds.

**Comment by slominskir** [comment]
Not sure why this is closed - when attempting to run on Windows using the Developer Notes / Run Tests procedure I still see:

`ImportError: DLL load failed while importing cimpl: The specified module could not be found.`

**Comment by edenhill** [comment]
@slominskir What wheel and python version?

**Comment by edenhill** [comment]
Please also provide the output from `pip install -v confluent_kafka`

**Comment by slominskir** [comment]
```
(.venv) C:\Users\ryans\IdeaProjects\confluent-kafka-python\.venv\Scripts>python -m site
sys.path = [
    'C:\\Users\\ryans\\IdeaProjects\\confluent-kafka-python\\.venv\\Scripts',
    'C:\\Users\\ryans\\IdeaProjects\\confluent-kafka-python\\.venv',
    'C:\\Users\\ryans\\IdeaProjects\\confluent-kafka-python\\.venv\\lib\\site-packages',
    'C:\\Users\\ryans\\AppData\\Local\\Programs\\Python\\Python39\\python39.zip',
    'C:\\Users\\ryans\\AppData\\Local\\Programs\\Python\\Python39\\DLLs',
    'C:\\Users\\ryans\\AppData\\Local\\Programs\\Python\\Python39\\lib',
    'C:\\Users\\ryans\\AppData\\Local\\Programs\\Python\\Python39',
    'C:\\Users\\ryans\\IdeaProjects\\confluent-kafka-python\\.venv\\lib\\site-packages\\confluent_kafka-1.6.1-py3.9-win-amd64.egg',
]
USER_BASE: 'C:\\Users\\ryans\\AppData\\Roaming\\Python' (doesn't exist)
USER_SITE: 'C:\\Users\\ryans\\AppData\\Roaming\\Python\\Python39\\site-packages' (doesn't exist)
ENABLE_USER_SITE: False
```

**Comment by edenhill** [comment]
Check if you have dumpbin.exe installed, then run `dumpbin /dependents  ..` on the following files in site-packages/confluent_kafka/...../..somewhere:
```
./confluent_kafka/cimpl.cp39-win_amd64.pyd
./confluent_kafka.libs/vcruntime140.dll
./confluent_kafka.libs/zlib-306d40ae.dll
./confluent_kafka.libs/librdkafka-04c02310.dll
./confluent_kafka.libs/libzstd-d6bdd78c.dll
./confluent_kafka.libs/msvcr120.dll

```

**Comment by slominskir** [comment]
There are several copies of the file _cimpl.cp39-win_amd64.pyd_, not sure which one is being used:

- **build dir:** C:\Users\ryans\IdeaProjects\confluent-kafka-python\build\lib.win-amd64-3.9\confluent_kafka\cimpl.cp39-win_amd64.pyd
- **venv:** C:\Users\ryans\IdeaProjects\confluent-kafka-python\.venv\Lib\site-packages\confluent_kafka-1.6.1-py3.9-win-amd64.egg\confluent_kafka\cimpl.cp39-win_amd64.pyd
- **interpreter  dir:** C:\Users\ryans\AppData\Local\Programs\Python\Python39\Lib\site-packages\confluent_kafka\cimpl.cp39-win_amd64.pyd

For **build dir** dumpbin shows:

```
C:\Users>dumpbin /dependents C:\Users\ryans\IdeaProjects\confluent-kafka-python\build\lib.win-amd64-3.9\confluent_kafka\cimpl.cp39-win_amd64.pydC:\Users\ryans\IdeaProjects\confluent-kafka-python\build\lib.win-amd64-3.9\confluent_kafka\cimpl.cp39-win_amd64.pydC:\Users\ryans\IdeaProjects\confluent-kafka-python\build\lib.win-amd64-3.9\confluent_kafka\cimpl.cp39-win_amd64.pyd
Microsoft (R) COFF/PE Dumper Version 14.28.29914.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file C:\Users\ryans\IdeaProjects\confluent-kafka-python\build\lib.win-amd64-3.9\confluent_kafka\cimpl.cp39-win_amd64.pydC:\Users\ryans\IdeaProjects\confluent-kafka-python\build\lib.win-amd64-3.9\confluent_kafka\cimpl.cp39-win_amd64.pydC:\Users\ryans\IdeaProjects\confluent-kafka-python\build\lib.win-amd64-3.9\confluent_kafka\cimpl.cp39-win_amd64.pyd
LINK : fatal error LNK1104: cannot open file 'C:\Users\ryans\IdeaProjects\confluent-kafka-python\build\lib.win-amd64-3.9\confluent_kafka\cimpl.cp39-win_amd64.pydC:\Users\ryans\IdeaProjects\confluent-kafka-python\build\lib.win-amd64-3.9\confluent_kafka\cimpl.cp39-win_amd64.pydC:\Users\ryans\IdeaProjects\confluent-kafka-python\build\lib.win-amd64-3.9\confluent_kafka\cimpl.cp39-win_amd64.pyd'
```

For both **venv dir** and **interpreter dir** dumpbin shows: 

```
C:\Users>dumpbin /dependents C:\Users\ryans\IdeaProjects\confluent-kafka-python\.venv\Lib\site-packages\confluent_kafka-1.6.1-py3.9-win-amd64.egg\confluent_kafka\cimpl.cp39-win_amd64.pyd
Microsoft (R) COFF/PE Dumper Version 14.28.29914.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file C:\Users\ryans\IdeaProjects\confluent-kafka-python\.venv\Lib\site-packages\confluent_kafka-1.6.1-py3.9-win-amd64.egg\confluent_kafka\cimpl.cp39-win_amd64.pyd

File Type: DLL

  Image has the following dependencies:

    librdkafka.dll
    python39.dll
    KERNEL32.dll
    VCRUNTIME140.dll
    api-ms-win-crt-stdio-l1-1-0.dll
    api-ms-win-crt-heap-l1-1-0.dll
    api-ms-win-crt-runtime-l1-1-0.dll
    api-ms-win-crt-string-l1-1-0.dll

  Summary

        D000 .data
        1000 .pdata
        5000 .rdata
        1000 .reloc
        1000 .rsrc
        E000 .text
```

**Comment by slominskir** [comment]
Not sure where any of the `confluent_kafka.libs/*` are.   Maybe they're missing?   Are they supposed to be under site-packages?

**Comment by slominskir** [comment]
If I manually download the wheel file from: https://www.wheelodex.org/projects/confluent-kafka/wheels/confluent_kafka-1.6.1-cp39-cp39-win_amd64.whl/ I can see the `confluent_kafka.libs` are inside.   However, I can't find the libs in my local install and trying to manually install the wheel doesn't work:

```
(.venv) C:\Users\ryans\IdeaProjects\confluent-kafka-python>pip install C:\Users\ryans\Downloads\confluent_kafka-1.6.1-cp39-cp39-win_amd64.whl
Requirement already satisfied: confluent-kafka==1.6.1 from file:///C:/Users/ryans/Downloads/confluent_kafka-1.6.1-cp39-cp39-win_amd64.whl in c:\users\ryans\ideaprojects\confluent-kafka-python\.venv\lib\site-packages\confluent_kafka-1.6.1-py3.9-win-amd64.egg (1.6.1)
```

I also noticed my install appears to be using an "egg" package format vs the expected wheel format:

 - **actual:** confluent_kafka-1.6.1-py3.9-win-amd64.egg
 - **expected:** confluent_kafka-1.6.1-cp39-cp39-win_amd64.whl

What could cause this?   Copy and pasting the confluent_kafka.libs directory into the _.venv/Lib/site-packages_ beside the _confluent_kafka-1.6.1-py3.9-win-amd64.egg_ doesn't appear to help either.

**Comment by edenhill** [comment]
Do `pip uninstall confluent_kafka` repeatedly until all installed versions are gone, then try to install the wheel manually

**Comment by slominskir** [comment]
That seems to have worked.   Had two versions (regular and egg) installed at the interpreter and one version (egg) installed in the virtual environment.   Not sure where the egg is coming from if that was removed from the build process and replaced with a wheel.

**Comment by slominskir** [comment]
Also noticed if I use `pip install -e .` the project is broken again.   How are developers supposed to make changes to source files and then run the unit tests?

**Comment by edenhill** [comment]
I typically do `python setup.py develop`  when in a devel virtualenvironment

**Comment by Reid00** [comment]
the same issue, don't know how to resolve, is there method?

**Comment by slominskir** [comment]
@Reid00 - I was able to manually install the wheel downloaded from here:  https://www.wheelodex.org/projects/confluent-kafka/wheels/confluent_kafka-1.6.1-cp39-cp39-win_amd64.whl/ and install it.   First I had to run:

```
pip uninstall confluent_kafka
```

Repeatedly as suggested by edenhill.   This allowed me to run the tests.   However, I never did figure out how to make changes in the source and then have them show up in the tests.   Neither `pip install -e` or  `python setup.py develop` worked for me.   So I'm very painfully relying on Travis CI for feedback on pull-requests.
